pmda/rmsf.py

```diff
@@ -0,0 +1,227 @@
+    with the averaged coordinates of the protein.  Now we need to fit the
+    whole trajectory to the reference by minimizing the RMSD. We use
+    :class:`MDAnalysis.analysis.align.AlignTraj`::
+
+        aligner = align.AlignTraj(u, reference, select="protein and name CA",
+                                  in_memory=True).run()
+
+    The trajectory is now fitted to the reference (the RMSD is stored as
+    `aligner.rmsd` for further inspection). Now we can calculate the RMSF::
+
+        from pmda.rmsf import RMSF
+        calphas = protein.select_atoms("name CA")
+        rmsfer = RMSF(calphas).run()
+
+    and plot::
+
+        import matplotlib.pyplot as plt
+        plt.plot(calphas.resnums, rmsfer.rmsf)
+
+    .. versionadded:: 0.3.0
+
+    """
+
+    def __init__(self, atomgroup):
+        u = atomgroup.universe
+        super(RMSF, self).__init__(u, (atomgroup, ))
+        self._atomgroup = atomgroup
+        self._top = u.filename
+        self._traj = u.trajectory.filename
+
+    def _prepare(self):
+        self.sumsquares = np.zeros((self._atomgroup.n_atoms, 3))
+        self.mean = self.sumsquares.copy()
+
+    def _single_frame(self, ts, atomgroups):
+        sumsquares = self.sumsquares
+        mean = self.mean
+        agroup = atomgroups[0]
+        return agroup, mean, sumsquares
+
+    def _conclude(self):
+        """
+        self._results : Array
+            (n_blocks x 2 x N x 3) array
+        """
+        n_blocks = len(self._results)
+        # serial case
+        if n_blocks == 1:
+            # get length of trajectory slice
+            k = len(self._blocks[0])
+            self.mean = self._results[0, 0]
+            self.sumsquares = self._results[0, 1]
+            self.rmsf = np.sqrt(self.sumsquares.sum(axis=1) / k)
+        # parallel case
+        else:
+            mean = self._results[:, 0]
+            sos = self._results[:, 1]
+            # create list of (timesteps, mean, sumsq tuples for each block
+            vals = []
+            for i in range(n_blocks):
+                vals.append((len(self._blocks[i]), mean[i], sos[i]))
+            # combine block results using fold method
+            results = fold_second_order_moments(vals)
+            self.totalts = results[0]
+            self.mean = results[1]
+            self.sumsquares = results[2]
+            self.rmsf = np.sqrt(self.sumsquares.sum(axis=1) / self.totalts)
+            self._negative_rmsf(self.rmsf)
+
+    @staticmethod
+    def _reduce(res, result_single_frame):
+        """
+        'sum' action for time series
+        """
+        atoms, mean, sumsq = result_single_frame
+        positions = atoms.positions.astype(np.float64)
+        # initial time step case
+        if isinstance(res, list) and len(res) == 0:
+            # set initial time step for each block to zero
+            k = 0
+            # reset sum of squares array for each block
+            sumsq = np.zeros_like((sumsq))
```

Comment:
@orbeckst Actually that's this line which fixed the bug. For `n_jobs=1 n_blocks=n` case, the `sumsq` would not be zero for the first frame in blocks with id larger than 1. So at here, we reset the `sumsq`with zero for the first frame for the block. 
I think `self.sumsquares` may be modified during the first block, and when the second block starts working, it will take the modified `self.sumsquares` as the initial `sumsq`.

Reply:
This is weird. Do we _know_ that this happens? Is this is what you'd _expect_ to happen, given the code?

@nawtrey Please add a comment saying why this reset has to be done. This is the kind of information we'll forget and then a year later we have to rediscover it.

Also, the double parentheses look weird in `zeros_like((sumsq))`. Maybe just `(sumsq)`? Btw, this will make a brand-new array and the old one will be garbage collected. If you really want to reset the array to zeros, I'd do one of
```python
sumsq *= 0
```
or
```python
sumsq[:] = 0
```
(the latter is marginally faster when I tested it)

Reply:
@orbeckst We did observe this directly through systematic testing of a short trajectory. It looks like the reason `mean` was not having a similar issue for `n_jobs=1` is that the mean is redefined every time a block starts the first frame, but the sum of squares is not. Adding this line resets the sum of squares the same way the mean is reset for the `k=0` case.

Reply:
I tried to make the change from

`sumsq = np.zeros_like(sumsq)`

to either of the options shown above, and both options fail test_RMSF_n_jobs() for the same cases as before (n_blocks = 2, 3, 4, 5 ,6, 7, 8, 9). The plot thickens.

Reply:
@nawtrey We can add another output to record `sumsq` for each frame in a test version. This would give us more details about this problem.

Reply:
For the `n_jobs=1` case, each consecutive block after the first was reading the prior block's sum of squares from self.sumsquares due to the  `_single_frame()` assignments:

`sumsquares = self.sumsquares`
`mean = self.mean`

These were changed to 

`sumsquares = np.zeros((self._atomgroup.n_atoms, 3))`
 `mean = sumsquares.copy()`

to resolve this issue. This now raises the question of whether or not it is necessary to make these assignments in `_single_frame()` in the first place. These could be placed into `_reduce()` for the `k=0` case which would remove the necessity to read them in from `_single_frame()`.